### DISCUSSION TITLE
P2P_MAX_DATA_SIZE

### DISCUSSION THREAD

**Discussion by imreburn** [bitcraze/discussions#919] [Q&A] [answered]
Hello,

This would be a quick question. According to the P2P API, the maximum payload size is currently set as `60`.(https://www.bitcraze.io/documentation/repository/crazyflie-firmware/master/functional-areas/p2p_api/)

I need to send data larger than this limit (~100 B). Is this `60` bytes limit a soft or hard limit? I think I can either change this macro (`P2P_MAX_DATA_SIZE`)  or divide the message into multiple packets. What would be side effects for either decision? I would appreciate if anyone can share experience with larger packet size. 

Thank you!

**Comment by ataffanel** [comment, marked as answer]
Hi,

It is technically a soft limit in the sense that the radio can send bigger packet, but it is on the hard-side of soft in regards to what would need to be changed in order to send bigger packet:

The nRF51 radio is currently setup in a mode compatible with Crazyradio PA's nRF24 chip. In this mode, the packet size field is 6 bit longs so it can encode packet size up to 63 bytes. Minus some housekeeping bytes we have the 60 Bytes MTU.

The main issue there, is that while this length field width can be changed only for sending P2P packet, it would need to be changed on receive mode as well making the Crazyflie incompatible with Crazyradio PA. One good news though is that Crazyradio 2.0 has a radio that will be able to work with it, as long as the firmware is modified for it of course.

So the procedure to increase the P2P MTU would be:
 - Change the MTU in the STM32
 - Change it in the nRF51
 - Make sure SYSLINK (the protocol between nrf51<->stm32) has a MTU big enough.
 - Change the nRF51 radio configuration (in esb.c)
 - Change the Crazyradio 2.0's radio configuration (no need to change MTU there if no 100bytes packet needs to be sent to the ground, bigger packet on the Crazyradio 2.0 requires to use the new unreleased radio protocol which is yet another work package ...).

So, in conclusion, it is possible but not trivial. It would be much easier to split the packet in 2 and send it in 2 times if possible. We are planing to work on new radio protocols with Crazyradio 2.0 that would allow this kind of use-case, but we will only have time to start working on it by the end of the year.

**Reply by imreburn** [reply]
Thank you so much for the answer and related information which is also useful and interesting. Since I operate CFs in decentralized, autonomous way (thank you for sharing ICRA demo, that code base helps a lot to me), I do not use Crazyradio most of the time. However, it still sounds that it would be simpler to divide packets into multiple ones since it can be achieved in App layer without modifying backbone.
